components/script/dom/globalscope.rs

```diff
@@ -2561,26 +2570,58 @@ impl GlobalScope {
                 let ar = enter_realm(&*self);
 
                 let _aes = AutoEntryScript::new(self);
-                let options =
-                    unsafe { CompileOptionsWrapper::new(*cx, filename.as_ptr(), line_number) };
-
-                debug!("evaluating Dom string");
-                let result = unsafe {
-                    Evaluate2(
-                        *cx,
-                        options.ptr,
-                        &mut transform_str_to_source_text(code),
-                        rval,
-                    )
-                };
 
-                if !result {
-                    debug!("error evaluating Dom string");
-                    unsafe { report_pending_exception(*cx, true, InRealm::Entered(&ar)) };
-                }
+                unsafe {
+                    let options = CompileOptionsWrapper::new(*cx, filename.as_ptr(), line_number);
+
+                    debug!("compiling Dom string");
+                    rooted!(in(*cx) let compiled_script =
+                        Compile1(
+                            *cx,
+                            options.ptr,
+                            &mut transform_str_to_source_text(code),
+                        )
+                    );
+
+                    if compiled_script.is_null() {
+                        debug!("error compiling Dom string");
+                        report_pending_exception(*cx, true, InRealm::Entered(&ar));
+                        return false;
+                    }
+
+                    // When `ScriptPrivate` for the compiled script is undefined,
+                    // we need to set it so that it can be used in dynamic import context.
+                    if GetScriptPrivate(*compiled_script).is_undefined() {
+                        let base_url = script_url.unwrap_or(self.api_base_url());
+
+                        debug!("Set script private for {}", base_url);
+
+                        let module_script_data = Box::new(ModuleScript::new(
+                            base_url,
+                            ScriptFetchOptions::default_classic_script(&self),
```

Comment:
I also noticed maybe we should use its host fetch options. Does it make sense to make `HTMLScriptElement` hold a `ScriptFetchOptions` so that it can be reused here when trying to evaluate the script?

Also, for cases like devtools, should we accept an Option here and fallback to default or just accept a `ScriptFetchOptions` 👀?

Reply:
I think storing it in HTMLScriotElement might be awkward, since it will either be an Option or a fake default value for part of it's existence. If we can store it in Script origin and ScriptContext while the fetching is happening, then we can create the ModuleScript with the real value. Does that make sense?

Reply:
> If we can store it in Script origin and ScriptContext while the fetching is happening, then we can create the ModuleScript with the real value.

ah! yes! this makes sense to me! I will try to fix it in this way! thanks!

Reply:
Hmm, while trying to add options to ScriptOrigin, I notice ScriptOrigin derives JSTraceable and MallocSizeOf traits which means we will need to make all properties of ScriptFetchOptions to derive these 2 traits?

If we try to pass `options` to all the `*_loaded` functions and pass it to the `execute` function, then we can reuse the options in `run_a_classic_script`. Does that make sense?

Reply:
If that makes it easier, then that sounds fine.

Reply:
hmmm... I found I also need to update a bunch of places in the codebase with passing options; maybe deriving the traits would be easier 😂

Reply:
hmm, no, while I saw this compilation error, I think it might be better to use the passing options way because I believe we want to use `JSTraceable` internally inside the script crate?

```error: cannot find derive macro `JSTraceable` in this scope
  --> components/net_traits/request.rs:37:37
   |
37 | #[derive(Clone, Debug, Deserialize, JSTraceable, MallocSizeOf, PartialEq, Serialize)]
   |                                     ^^^^^^^^^^^
```

Will use the passing options way to fix this!

Reply:
The other alternative is to add a new entry with `unsafe_no_jsmanaged_fields!` in components/script/dom/bindings/trace.rs.

Reply:
I fixed this with 3 separated commits.

- 870692c0aa55 : Make `script_base_url` non-optional
- 8c1eba6cdbd8 : Make `ScriptOrigin` to hold its fetch options
- f300f22396aa58 : Use fetch options from ScriptOrigin

Hope this will make review easier 🙏

Reply:
Btw, using `unsafe_no_jsmanaged_fields!` in this case makes sense to me because I think we might not need to trace those options and it works fine! Thanks for the tip!